// aifactory/aifactory.financials.reconciliation-1.0.0.js
/*
	AI factory; Financials
	See README.md
*/

var mydigitalstructure = require('mydigitalstructure')
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '1.0.0',

	init: function (param)
	{
        mydigitalstructure.add(
        {
            name: 'ai-financials-reconciliation-bank-account-latest',
            code: function (param, response)
            {
                var event = mydigitalstructure.get({scope: '_event'});

                if (response == undefined)
                {
                    var filters =
                    [
                        {
                            field: 'status',
                            comparison: 'EQUAL_TO',
                            value: '1'
                        },
                        {
                            field: 'bankaccount',
                            comparison: 'EQUAL_TO',
                            value: event.bankAccountID
                        }
                    ];

                    mydigitalstructure.cloud.search(
                    {
                        object: 'financial_reconciliation',
                        fields:
                        [
                            {name: 'statementdate'},
                            {name: 'previousbalance'},
                            {name: 'bankaccount'}, {name: 'bankaccounttext'},
                            {name: 'status'},  {name: 'statustext'},
                            {name: 'statementbalancecredits'},
                            {name: 'statementbalancedebits'},
                            {name: 'guid'}
                        ],
                        filters: filters,
                        sorts: 
                        [
                            {
                                field: 'statementdate',
                                direction: 'desc'
                            }
                        ],
                        rows: 1,
                        callback: 'ai-financials-reconciliation-bank-account-latest',
                        callbackParam: param
                    });
                }
                else
                {
                    var reconciliation = mydigitalstructure.set(
                    {
                        scope: 'ai-financials-reconciliation-bank-account',
                        context: 'latest',
                        value: response.data.rows
                    });

                    mydigitalstructure._util.message(reconciliation, 'ai-financials-reconciliation-bank-account-latest');

                    mydigitalstructure._util.onComplete(param);
                }
            }
        });
    }
}
