// aifactory/aifactory.financials-1.0.0.js
/*
	AI factory; Financials
	See README.md
*/

var mydigitalstructure = require('mydigitalstructure')
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '1.0.0',

	init: function (param)
	{
		// MAPPINGS

		mydigitalstructure.add(
		{
			name: 'ai-financials-bank-account-apply-mappings',
			code: function (param, response)
			{
				mydigitalstructure.invoke('ai-financials-util-bank-accounts-mappings',
				{
					onComplete: 'ai-financials-bank-account-apply-mappings-process'
				});
			}
		});

		mydigitalstructure.add(
		{
			name: 'ai-financials-bank-account-apply-mappings-process',
			code: function (param, response)
			{
                
				mydigitalstructure.invoke('ai-financials-util-bank-accounts-mappings-apply',
				{
					onComplete: 'ai-financials-bank-account-apply-mappings-save'
				});
			}
		});

        mydigitalstructure.add(
        {
            name: 'ai-financials-bank-account-apply-mappings-save',
            code: function (param, response)
            {
                mydigitalstructure.invoke('ai-financials-util-bank-accounts-matched-transactions-save',
                {
                    onComplete: 'ai-financials-bank-account-apply-mappings-complete'
                });
            }
        });

		mydigitalstructure.add(
		{
			name: 'ai-financials-bank-account-apply-mappings-complete',
			code: function (param, response)
			{
				var summary = mydigitalstructure.get(
				{
					scope: 'ai-financials-bank-account-apply-mappings',
					context: 'summary'
				});
		
				mydigitalstructure.invoke('util-end', summary)
			}
		});
	
		mydigitalstructure.add(
		{
			name: 'ai-financials-bank-account-transactions-status',
			code: function (param, response)
			{
				mydigitalstructure.invoke('ai-financials-util-bank-accounts-transactions-status',
				{
					onComplete: 'ai-financials-bank-reconciliation-mappings'
				});
			}
		});
	}
}
